homeassistant/components/izone/discovery.py

```diff
@@ -0,0 +1,132 @@
+"""Internal discovery service for  iZone AC."""
+
+import logging
+from asyncio import Event
+from typing import Dict
+
+import pizone
+
+from homeassistant.const import CONF_EXCLUDE, EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_send, async_dispatcher_connect)
+
+from .climate import init_controller
+from .const import (
+    DATA_ADD_ENTRIES, DATA_CONFIG, DATA_DISCOVERY_SERVICE,
+    DISPATCH_CONTROLLER_DISCOVERED,
+    DISPATCH_CONTROLLER_DISCONNECTED, DISPATCH_CONTROLLER_RECONNECTED,
+    DISPATCH_CONTROLLER_UPDATE, DISPATCH_ZONE_UPDATE)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DiscoveryService(pizone.Listener):
+    """Discovery data and interfacing with pizone library."""
+
+    def __init__(self, hass):
+        """Initialise discovery service."""
+        super().__init__()
+        self.hass = hass
+        self.controllers = {}  # type: Dict[str, pizone.Controller]
+        self.controller_ready = Event()
+
+        self.pi_disco = None
+        self.stop_listener = None
+
+        async def _controller_discovered(ctrl: pizone.Controller):
+            assert ctrl.device_uid not in self.controllers, \
+                "discovered device that already exists"
+            _LOGGER.debug("Controller discovered uid=%s", ctrl.device_uid)
+
+            conf = self.hass.data.get(DATA_CONFIG)  # type: ConfigType
+
+            # Filter out any entities excluded in the config file
+            if conf and ctrl.device_uid in conf[CONF_EXCLUDE]:
+                _LOGGER.info(
+                    "Controller UID=%s ignored as excluded.",
+                    ctrl.device_uid)
+                return
+
+            self.controllers[ctrl.device_uid] = ctrl
+            self.controller_ready.set()
+
+            # This will be present if the component is configured.
+            # otherwise init_controller will be called when the config entry
+            # is created.
+            async_add_entries = self.hass.data.get(DATA_ADD_ENTRIES)
+            if async_add_entries:
+                init_controller(ctrl, async_add_entries)
```

Comment:
Instead connect `init_controller` to a signal and signal the climate platform to add an entity.

We don't need to store `async_add_entities` in hass.data then either.

We shouldn't import functions from an entity platform but keep the platform contained.

See eg:
https://github.com/home-assistant/home-assistant/blob/7ff7c7b9f5b7fcac5483cb975fe88bd747bbe4e8/homeassistant/components/zha/fan.py#L55-L84

Reply:
Fixed.